### PR TITLE
networkd: Introduce geneve GEneric NEtwork Virtualization Encapsulation

### PR DESCRIPTION
**Pull request by ssahani**
This work enables cration of geneve tunnel.

### CODE REVIEW

File: src/network/netdev/netdev-gperf.gperf [hunk 1]
```diff
@@ -79,6 +80,15 @@ VXLAN.MaximumFDBEntries,     config_parse_unsigned,                0,
 VXLAN.PortRange,             config_parse_port_range,              0,                             0
 VXLAN.DestinationPort,       config_parse_destination_port,        0,                             offsetof(VxLan, dest_port)
 VXLAN.FlowLabel,             config_parse_flow_label,              0,                             0
+GENEVE.Id,                   config_parse_uint32,                  0,                             offsetof(Geneve, id)
+GENEVE.Remote,               config_parse_geneve_address,          0,                             offsetof(Geneve, remote)
+GENEVE.TOS,                  config_parse_unsigned,                0,                             offsetof(Geneve, tos)
+GENEVE.TTL,                  config_parse_unsigned,                0,                             offsetof(Geneve, ttl)
```
Comment (poettering) [inline src/network/netdev/netdev-gperf.gperf]: the man page you added says the range for this is 1-155 (is that true, isn't that a typo? should that be 255?). If so, then please add config_parse_geneve_ttl() or so, which ensures the value passed is in the right range.

File: src/network/netdev/geneve.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+        r = parse_ip_port(rvalue, &port);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to parse GENEVE destination port '%s'.", rvalue);
+                return 0;
+        }
+
+        v->dest_port = port;
+
+        return 0;
```
Comment (poettering) [inline src/network/netdev/geneve.c:333]: We should really generalize this into a generic config_parse_ip_port() or so which is added to conf-parser.h, and then used wherever we parse IP ports, which I figure is a non-zero number of places already. Any chance you can make this happen and rework the users of parse_ip_port()?

File: src/network/netdev/geneve.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#include "networkd-link.h"
+#include "networkd-network.h"
+#include "netdev.h"
+
+#define GENEVE_VID_MAX (1u << 24) - 1
+
+struct Geneve {
+        NetDev meta;
+
+        uint64_t id;
```
Comment (poettering) [inline src/network/netdev/geneve.h]: hu? is this 32bit or 64bit? You aprse this with config_parse_uint32(), but define the type to be uint64_t... that looks very wrong..

File: src/network/netdev/geneve.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        int remote_family;
+
+        union in_addr_union remote;
+
+        unsigned tos;
+        unsigned ttl;
```
Comment (poettering) [inline src/network/netdev/geneve.h]: the kernel accepts this as u8, hence this should have the type uint8_t here, too

File: man/systemd.netdev.xml [hunk 5]
```diff
@@ -615,6 +618,69 @@
...
+      <varlistentry>
+        <term><varname>Remote=</varname></term>
+        <listitem>
+          <para>Specifies the unicast destination IP address to use in outgoing packets.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>TOS=</varname></term>
+        <listitem>
+          <para>Specifies the TOS value to use in outgoing packets.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: is this an integer? if so, say so. what's the valid range for it?

File: src/network/netdev/netdev-gperf.gperf [hunk 6]
```diff
@@ -79,6 +80,15 @@ VXLAN.MaximumFDBEntries,     config_parse_unsigned,                0,
 VXLAN.PortRange,             config_parse_port_range,              0,                             0
 VXLAN.DestinationPort,       config_parse_destination_port,        0,                             offsetof(VxLan, dest_port)
 VXLAN.FlowLabel,             config_parse_flow_label,              0,                             0
+GENEVE.Id,                   config_parse_uint32,                  0,                             offsetof(Geneve, id)
+GENEVE.Remote,               config_parse_geneve_address,          0,                             offsetof(Geneve, remote)
+GENEVE.TOS,                  config_parse_unsigned,                0,                             offsetof(Geneve, tos)
```
Comment (poettering) [inline src/network/netdev/netdev-gperf.gperf]: please use a proper parser function for this, and generate a parsing error log msg for invalid values

File: src/network/netdev/geneve.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        unsigned flow_label;
```
Comment (poettering) [inline src/network/netdev/geneve.h]: the kernel wants this as uint32_t... please use the right type here too

File: src/network/netdev/geneve.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+struct Geneve {
+        NetDev meta;
+
+        uint64_t id;
+
+        int remote_family;
+
+        union in_addr_union remote;
+
+        unsigned tos;
```
Comment (poettering) [inline src/network/netdev/geneve.h]: please use uint8_t here, if that's what the kernel expects

File: src/network/netdev/geneve.c (hunk 2 above)
Comment (ssahani) [reply src/network/netdev/geneve.c:333]: Yes I am thinking of making it a separate patch.

File: src/network/netdev/geneve.c [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+        assert(rvalue);
+        assert(data);
+
+        r = safe_atou8(rvalue, &f);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to parse Geneve TOS '%s'.", rvalue);
+                return 0;
+        }
+
+        v->tos = f;
```
Comment (poettering) [inline src/network/netdev/geneve.c:229]: hmm, this function is entirely generic, and accepts all kinds of uint8_t, hence I think we should just introduce config_parse_uint8() or so as part of conf-parser.[ch], similar in style to the existing config_parse_uint16(), config_parse_uint32() and we have and use that.

**Comment by poettering** [comment]
OK, looks good to me. But I'd really appreciate a follow-up commit that adds generic config_parse_uint8() and config_prase_ip_port() functions and uses then for this stuff and elsewhere, see comments.

**Pull request merged**